packages/thirdweb/src/react/web/hooks/useSendTransaction.tsx

```diff
@@ -0,0 +1,207 @@
+import { useContext, useEffect, useState } from "react";
+import type { ThirdwebClient } from "../../../client/client.js";
+import { useActiveAccount } from "../../../exports/react-native.js";
+import {
+  type WaitForReceiptOptions,
+  estimateGasCost,
+} from "../../../exports/transaction.js";
+import { resolvePromisedValue, toEther } from "../../../exports/utils.js";
+import { getWalletBalance } from "../../../exports/wallets.js";
+import type { PreparedTransaction } from "../../../transaction/prepare-transaction.js";
+import { useSendTransactionCore } from "../../core/hooks/contract/useSendTransaction.js";
+import { SetRootElementContext } from "../../core/providers/RootElementContext.js";
+import {
+  type SupportedTokens,
+  defaultTokens,
+} from "../ui/ConnectWallet/defaultTokens.js";
+import { getConnectLocale } from "../ui/ConnectWallet/locale/getConnectLocale.js";
+import type { ConnectLocale } from "../ui/ConnectWallet/locale/types.js";
+import { BuyScreen } from "../ui/ConnectWallet/screens/Buy/SwapScreen.js";
+import { Modal } from "../ui/components/Modal.js";
+import { Spacer } from "../ui/components/Spacer.js";
+import { Container, ModalHeader } from "../ui/components/basic.js";
+import { Text } from "../ui/components/text.js";
+import { CustomThemeProvider } from "../ui/design-system/CustomThemeProvider.js";
+import type { Theme } from "../ui/design-system/index.js";
+import type { LocaleId } from "../ui/types.js";
+import { LoadingScreen } from "../wallets/shared/LoadingScreen.js";
+
+type SendTransactionConfig = {
+  buyModal: {
+    locale: LocaleId;
+    supportedTokens: SupportedTokens;
+    theme: Theme | "light" | "dark";
+  };
+};
+
+/**
+ * A hook to send a transaction.
+ * @returns A mutation object to send a transaction.
+ * @example
+ * ```jsx
+ * import { useSendTransaction } from "thirdweb/react";
+ * const { mutate: sendTx, data: transactionResult } = useSendTransaction();
+ *
+ * // later
+ * sendTx(tx);
+ * ```
+ *
+ * @transaction
+ */
+export function useSendTransaction(config?: SendTransactionConfig) {
+  const sendTxCore = useSendTransactionCore();
+  const setRootEl = useContext(SetRootElementContext);
+  const account = useActiveAccount();
+
+  return useMutation({
+    mutationFn: async (tx: PreparedTransaction) => {
+      if (!account) {
+        throw new Error("No active account");
+      }
+
+      const address = account.address;
+
+      return new Promise<WaitForReceiptOptions>((resolve, reject) => {
+        const sendTx = async () => {
+          try {
+            const res = await sendTxCore.mutateAsync(tx);
+            resolve(res);
+          } catch (e) {
+            reject(e);
+          }
+        };
+
+        (async () => {
+          try {
+            const walletBalance = await getWalletBalance({
+              address: address,
+              chain: tx.chain,
+              client: tx.client,
+            });
+
+            const txValue = await resolvePromisedValue(tx.value);
```

Comment:
I think its safer if you get the `tx.value` *after* u `estimateGasCost()` because it may get set by the async params on the transaction

however you can get wallet balance and the gas cost in parallel and then resolve the tx.value after that

so Promise.all(getwalletBalance, estimateGasCost) -> resolvePromisedValue(tx.value)

Reply:
Thanks!